Mathlib/Order/IsWellOrderLimitElement.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Data.Nat.Order.Basic
+import Mathlib.Order.WellFounded
+
+/-!
+# Limit elements in well-ordered types
+
+This file introduces two main definitions:
+- `wellOrderSucc a`: the successor of an element `a` in a well-ordered type
+- the typeclass `IsWellOrderLimitElement a` which asserts that an element `a` (in a
+well-ordered type) is neither a successor nor the smallest element, i.e. `a` is a limit element
+
+Then, the lemma `eq_bot_or_eq_succ_or_isWellOrderLimitElement` shows that an element
+in a well-ordered type is either `⊥`, a successor, or a limit element.
+
+-/
+
+variable {α : Type*} [LinearOrder α] [IsWellOrder α (· < ·)]
+
+/-- Given an element `a : α` in a well ordered set, this is the successor of `a`,
+i.e. the smallest element stricly greater than `a` if it exists (or `a` itself otherwise). -/
+noncomputable def wellOrderSucc (a : α) : α :=
+  (@IsWellFounded.wf α (· < ·)).succ a
```

Comment:
```suggestion
  (IsWellFounded.wf (r := (· < ·))).succ a
```
Is it a good idea to avoid `@` if we can? Somehow I always think of `@` as a code smell.

Reply:
Yes, I completely agree!